lib/modem_info/modem_info.c

```diff
@@ -0,0 +1,412 @@
+	.param_count = MI_UICC_PARAM_COUNT,
+	.data_type = MODEM_DATA_TYPE_SHORT,
+};
+
+static struct modem_info_data battery_data = {
+	.cmd = AT_CMD_VBAT,
+	.valid_param = MI_VBAT_VALID_PARAM,
+	.param_count = MI_VBAT_PARAM_COUNT,
+	.data_type = MODEM_DATA_TYPE_SHORT,
+};
+
+static struct modem_info_data temp_data = {
+	.cmd = AT_CMD_TEMP,
+	.valid_param = MI_TEMP_VALID_PARAM,
+	.param_count = MI_TEMP_PARAM_COUNT,
+	.data_type = MODEM_DATA_TYPE_SHORT,
+};
+
+static struct modem_info_data fw_data = {
+	.cmd = AT_CMD_FW_VERSION,
+	.valid_param = MI_FW_VALID_PARAM,
+	.param_count = MI_FW_PARAM_COUNT,
+	.data_type = MODEM_DATA_TYPE_STRING,
+};
+
+static struct modem_info_data *modem_data[] = {
+	[MODEM_STATUS_RSRP] = &rsrp_data,
+	[MODEM_STATUS_BAND] = &band_data,
+	[MODEM_STATUS_MODE] = &mode_data,
+	[MODEM_STATUS_OPERATOR] = &operator_data,
+	[MODEM_STATUS_CELLID] = &cellid_data,
+	[MODEM_STATUS_IP_ADDRESS] = &ip_data,
+	[MODEM_STATUS_UICC] = &uicc_data,
+	[MODEM_STATUS_BATTERY] = &battery_data,
+	[MODEM_STATUS_TEMP] = &temp_data,
+	[MODEM_STATUS_FW_VERSION] = &fw_data,
+};
+
+static const char * const modem_data_name[] = {
+	[MODEM_STATUS_RSRP] = "RSRP",
+	[MODEM_STATUS_BAND] = "BAND",
+	[MODEM_STATUS_MODE] = "MODE",
+	[MODEM_STATUS_OPERATOR] = "OPERATOR",
+	[MODEM_STATUS_CELLID] = "CELLID",
+	[MODEM_STATUS_IP_ADDRESS] = "IP ADDRESS",
+	[MODEM_STATUS_UICC] = "SIM",
+	[MODEM_STATUS_BATTERY] = "BATTERY",
+	[MODEM_STATUS_TEMP] = "TEMP",
+	[MODEM_STATUS_FW_VERSION] = "FW",
+};
+
+static rsrp_cb_t modem_info_rsrp_cb;
+
+static struct at_param_list m_param_list;
+static struct k_thread socket_thread;
+static K_THREAD_STACK_DEFINE(socket_thread_stack, THREAD_STACK_SIZE);
+static struct k_mutex socket_mutex;
+static int at_socket_fd = INVALID_DESCRIPTOR;
+static struct pollfd fds[1];
+static int nfds;
+
+static int at_cmd_send(int fd, const char *cmd, size_t size, char *resp_buffer)
+{
+	int len;
+
+	k_mutex_lock(&socket_mutex, K_FOREVER);
+
+	LOG_DBG("send: %s\n", cmd);
+	len = send(fd, cmd, size, 0);
+	LOG_DBG("len: %d\n", len);
+	if (len != size) {
+		LOG_DBG("send: failed\n");
+		return -EIO;
+	}
+
+	if (resp_buffer != NULL) {
+		len = recv(fd, resp_buffer, CONFIG_MODEM_INFO_BUFFER_SIZE, 0);
+	}
+
+	if (len < 0) {
+		return -EFAULT;
+	}
+
+	k_mutex_unlock(&socket_mutex);
+
+	return 0;
+}
+
+static bool is_cesq_notification(char *buf, size_t len)
+{
+	if ((len < MI_CMD_SIZE(AT_CMD_CESQ_RESP)) ||
+	    (buf[4] != AT_CMD_CESQ_RESP[4])) {
+		return false;
+	}
+
+	return strstr(buf, AT_CMD_CESQ_RESP) ? true : false;
+}
+
+enum modem_data_type modem_info_get_type(enum modem_status status)
+{
+	__ASSERT(status < MODEM_STATUS_COUNT, "Invalid argument.");
+
+	return modem_data[status]->data_type;
+}
+
+int modem_info_get_name(enum modem_status status, char *data_name)
+{
+	int len;
+
+	if (data_name == NULL) {
+		return -EINVAL;
+	}
+
+	memcpy(data_name,
+		modem_data_name[status],
+		strlen(modem_data_name[status]));
+
+	len = strlen(modem_data_name[status]);
```

Comment:
It is possible by avoiding `strlen` by either implementing a clever x-macro or by simply defining the string length. I am hesitant to go for either strategy since a macro would damage the readability (and this is a public api) whilst having length defines is not that pretty in the long run in my opinion. 

I think using `strlen` here is non-optimal but is ok for this use case, as the strings themselves are short and the modem informer is not expected to execute this part of the code often (actually only on startup). 

Either way, I've moved line 240 above the `memcpy` of course, so we don't call `strlen` twice at least. :)